### PR TITLE
fix(license-detection): deterministic candidate iteration so deadline-truncated scans are reproducible

### LINKED ISSUE #1017: fix(license-detection): non-deterministic scan output when the matching deadline fires during candidate selection
## Confirmed bug (reproducibility)

`find_set_candidates` (`src/license_detection/seq_match/candidates/mod.rs:369-381`) builds `candidate_rids` as a `std::collections::HashSet<RuleId>` (SipHash, **iteration order randomized per process run**), then iterates it and, when the per-file matching deadline fires, `return`s the **partial subset collected so far** (line 379-380):

```rust
let candidate_rids: HashSet<RuleId> = ...collect();
for (rid_index, rid) in candidate_rids.into_iter().enumerate() {
    if rid_index.is_multiple_of(128) && deadline_exceeded(deadline) {
        return candidates; // partial, random-ordered subset
    }
    ...
}
```

Downstream sorting IS a total order (`compare_candidate_rank` ends with `.then_with(|| rid.cmp(&other_rid))`, and the final `b.cmp(a)`), so **when the full candidate set is collected the output is deterministic**. But when the deadline truncates the loop, a *different random subset* of candidate rules is collected each run → different shortlist → **different license detections run-to-run for the same input**.

### Impact
Non-reproducible scan output for license texts large/complex enough to trip the matching deadline. Invisible on normal corpora (the deadline never fires on small files), which is why extensive before/after byte-identical diffs (e.g. on apache/airflow) never surfaced it.

### Fix
Make the candidate iteration order deterministic so a deadline-truncated run returns a **deterministic prefix**: collect `candidate_rids` into a `Vec` and sort by `RuleId` (or use a `BTreeSet<RuleId>`) before the deadline-bounded loop. Cheap relative to the matching work. The other deadline early-exit (in `rescore_candidates_with_multisets`) operates on an already-ordered `Vec` and is unaffected.

### Test
Add a test that forces a tiny/zero deadline so the early-exit triggers, and asserts the produced candidates/detections are identical across repeated runs (and stable across process invocations). Note: output on deadline-hit inputs will change from the old arbitrary result to the new deterministic one — that is the intended fix; any golden shift must be a confirmed deadline-hit fixture.

### PR DESCRIPTION
**Pull request by mstykow**
## Summary

- `find_set_candidates` collected candidate rule IDs into a `HashSet<RuleId>`, whose SipHash iteration order is randomized per process run. When the per-file matching deadline fired, the function returned the partial subset collected so far. Because the iteration order was random, a deadline-truncated run collected a different arbitrary subset each run, producing non-deterministic (non-reproducible) license detections for inputs large enough to trip the deadline.
- Fix: collect the candidate rule IDs into a `Vec` and `sort_unstable()` by `RuleId` before the deadline-bounded loop, so a deadline-truncated run returns a deterministic prefix ordered by `RuleId`. The downstream sort/scoring is unchanged (it was already a total order via the `rid` tiebreak), and `rescore_candidates_with_multisets` already iterates an ordered `Vec`, so it needs no change.

## Issues

- Closes: #1017

## Scope and exclusions

- Included: deterministic candidate iteration in `find_set_candidates`; a regression test asserting deterministic full and deadline-truncated candidate selection.
- Explicit exclusions: no change to scoring, ranking, dedupe, or the downstream sorts. No golden fixtures regenerated (none shifted).

## How to verify

- `cargo test -p provenant-cli --lib license_detection::seq_match::candidates` exercises the new regression test, which asserts candidate iteration order is sorted by `RuleId` and byte-identical across repeated calls, including a run with an already-elapsed deadline that forces the early-exit branch.

## Expected-output fixture changes

- Files changed: none. The golden suite (`cargo test --test license_detection_golden --features golden-tests`) passes with 0 diffs because the golden fixtures are small and never trip the matching deadline.

Co-Authored-By: Claude Fable 5 <noreply@anthropic.com>

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/license_detection/seq_match/candidates/mod.rs [hunk 1]
```diff
@@ -366,13 +366,22 @@ fn find_set_candidates<'a>(
...
         .filter_map(|tid| index.rids_by_high_tid.get(&TokenId::new(tid)))
         .flat_map(|rids| rids.iter().copied())
         .collect();
 
+    // Sort the candidate rule IDs so the deadline-bounded loop below visits them
+    // in a deterministic order. `HashSet` iteration order is randomized per
+    // process run, so a deadline-truncated run would otherwise collect a
+    // different arbitrary subset of candidates each run, producing
+    // non-reproducible license detections on inputs large enough to trip the
+    // deadline.
```
[the remark this answers is not in the document]
Comment (mstykow) [reply src/license_detection/seq_match/candidates/mod.rs]: Done — trimmed to a single line: `// Sort so a deadline-truncated run returns a deterministic prefix rather than an arbitrary HashSet-ordered subset (issue #1017).`

File: src/license_detection/seq_match/candidates/mod.rs [hunk 2]
```diff
@@ -1063,4 +1072,178 @@ mod tests {
...
+        }
+
+        // A deadline that has already elapsed forces the early-exit branch. The
+        // truncated result must also be deterministic across repeated calls
+        // (here it is an empty deterministic prefix, since the deadline fires at
+        // the first checkpoint).
+        let elapsed = Instant::now()
+            .checked_sub(std::time::Duration::from_secs(60))
+            .expect("instant in the past");
+        let truncated = rid_sequence(&find_set_candidates(
```
[the remark this answers is not in the document]
Comment (mstykow) [reply src/license_detection/seq_match/candidates/mod.rs:1148]: Fixed. The test now inserts 200 rules so the candidate set spans more than one checkpoint (>128), and asserts (a) the full order is sorted, (b) it is byte-identical across repeated calls, and (c) a deadline-truncated run returns a deterministic prefix of the full ordering (`truncated == baseline[..truncated.len()]`). Since the deadline check only truncates the already-sorted Vec, that prefix invariant is what guarantees reproducibility regardless of exactly when the deadline fires — forcing a precise mid-loop wall-clock truncation would be timing-flaky, so the prefix assertion captures the property directly.

File: src/license_detection/seq_match/candidates/mod.rs [hunk 3]
```diff
@@ -1063,4 +1072,178 @@ mod tests {
             "Python final candidate tuple ordering falls back to higher rid after equal scores"
         );
     }
+
+    /// Inserts a minimal whitespace-tokenized rule into a test index, mirroring
+    /// the helper used by the `seq_match` module tests, and returns its `RuleId`.
+    fn insert_test_rule(index: &mut LicenseIndex, text: &str, expression: &str) -> RuleId {
```
[the remark this answers is not in the document]
Comment (mstykow) [reply src/license_detection/seq_match/candidates/mod.rs]: Done — hoisted the shared helper into `test_utils::add_text_rule` (cfg(test), pub(crate)). `seq_match::add_test_rule` now delegates to it, and the candidate test uses it too, so there is a single copy.
